pkg/server/event_handler.go

```diff
@@ -234,33 +240,57 @@ func (h *EventHandler[T, P]) HTTPHandler() http.Handler {
 // [GCS notification]: https://cloud.google.com/storage/docs/pubsub-notifications#format
 func (h *EventHandler[T, P]) Handle(ctx context.Context, m pubsub.Message) error {
 	logger := logging.FromContext(ctx)
+
+	eventBytes, processErr := h.generatePmapEventBytes(ctx, m)
+
+	attr := map[string]string{}
+
+	if processErr != nil {
+		if errors.Is(processErr, pmaperrors.ErrPubsubRetryable) {
+			return processErr
+		}
+		attr[PubSubAttrKeyNonRetryableErr] = processErr.Error()
+		logger.Errorw(processErr.Error(), "bucketId", m.Attributes["bucketId"], "objectId", m.Attributes["objectId"])
+		if err := h.failureMessenger.Send(ctx, eventBytes, attr); err != nil {
+			return fmt.Errorf("failed to send failure event downstream: %w", err)
+		}
+		return nil
+	}
+	if err := h.successMessenger.Send(ctx, eventBytes, attr); err != nil {
+		return fmt.Errorf("failed to send succuss event downstream: %w", err)
+	}
+	return nil
+}
+
+func (h *EventHandler[T, P]) generatePmapEventBytes(ctx context.Context, m pubsub.Message) ([]byte, error) {
+	// TODO(#20): The currently logic is to treat all errors that are not caused by processor as retryable.
+	// and all errors that are caused by processor as non-retryable.
+	// Therefore, we intend to keep this logic here, and do a more detailed differentiate later.
+
 	// Get the GCS object as a proto message given GCS notification information.
 	p, err := h.getGCSObjectProto(ctx, m.Attributes)
 	if err != nil {
-		return fmt.Errorf("failed to get GCS object: %w", err)
+		return nil, fmt.Errorf("%w: failed to get GCS object: %w", pmaperrors.ErrPubsubRetryable, err)
 	}
 
-	// TODO(#20): we need to have a way to differentiate retryable err vs. not.
-	// For non-retryable error, we need to have them enter a different BQ table per design.
-	// Currently all error events are sent to downstream if failureMessenger is provided
-	// including those retried events.
 	var processErr error
 	for _, processor := range h.processors {
 		if err := processor.Process(ctx, p); err != nil {
-			processErr = fmt.Errorf("failed to process object: %w", err)
+			processErr = fmt.Errorf("%w: failed to process object: %w", pmaperrors.ErrPubsubNonRetryable, err)
 			break
 		}
 	}
+
 	payload, err := anypb.New(p)
 	if err != nil {
-		return fmt.Errorf("failed to convert object to pmap event payload: %w", err)
+		return nil, fmt.Errorf("%w: failed to convert object to pmap event payload: %w", pmaperrors.ErrPubsubRetryable, err)
```

Comment:
We should consider join the error with the `processErr`

Reply:
the current behavior treat this as retryable, appending it to processErr would make it non-retryable and change the behavior. I think this should be included in TODO(#20)?